composer/trainer/trainer.py

```diff
@@ -997,6 +990,72 @@ def _train_loop(self) -> None:
                                                                                    event=Event.EPOCH_END):
                 self._checkpoint_saver.save_checkpoint(state=self.state, seed=self._seed, device=self._device)
 
+    def _handle_cuda_oom(self, e: RuntimeError):
+        """Handles CUDA Out of Memory and rescales if using adaptive grad_accum."""
+        if self.adaptive_grad_accum and "CUDA out of memory" in str(e):
+            # Raise runtime error if training 1 sample at a time still resulted in CUDA out of memory
+            if self.state.grad_accum == self.state.batch_num_samples:
+                raise RuntimeError("CUDA out of memory. Train loop failed with an internal microbatch of size 1")
+            else:
+                self.state.grad_accum = min(2 * self.state.grad_accum, self.state.batch_num_samples)
```

Comment:
Hm... 2 questions:
1) Why is it a problem if different nodes have different sized grad accum?
2) If we set it in a finally block, could this lead to slower speeds? Since now you're forcing sync in between batches